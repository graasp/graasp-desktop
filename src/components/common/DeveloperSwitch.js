import React, { Component } from 'react';
import FormControl from '@material-ui/core/FormControl';
import { withStyles } from '@material-ui/core/styles';
import { withTranslation } from 'react-i18next';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Switch from '@material-ui/core/Switch';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import { getDeveloperMode, setDeveloperMode } from '../../actions';
import Loader from './Loader';

const styles = theme => ({
  formControl: {
    margin: theme.spacing(),
    minWidth: 120,
  },
});

export class DeveloperSwitch extends Component {
  static propTypes = {
    developerMode: PropTypes.bool.isRequired,
    activity: PropTypes.bool.isRequired,
    t: PropTypes.func.isRequired,
    dispatchGetDeveloperMode: PropTypes.func.isRequired,
    dispatchSetDeveloperMode: PropTypes.func.isRequired,
    classes: PropTypes.shape({
      formControl: PropTypes.string.isRequired,
    }).isRequired,
  };

  constructor(props) {
    super(props);
    const { dispatchGetDeveloperMode } = this.props;
    dispatchGetDeveloperMode();
  }

  handleChange = async ({ target }) => {
    const { dispatchSetDeveloperMode } = this.props;
    const { checked } = target;
    dispatchSetDeveloperMode(checked);
  };

  render() {
    const { classes, t, developerMode, activity } = this.props;

    if (activity) {
      return <Loader />;
    }

    const control = (
      <Switch
        checked={developerMode}
        onChange={this.handleChange}
        value={developerMode}
        color="primary"
      />
    );

    return (
      <FormControl className={classes.formControl}>
        <FormControlLabel control={control} label={t('Developer Mode')} />
      </FormControl>
    );
  }
}

const mapStateToProps = ({ User }) => ({
  developerMode: User.getIn(['current', 'developerMode']),
  activity: Boolean(User.getIn(['current', 'activity']).size),
});

const mapDispatchToProps = {
  dispatchGetDeveloperMode: getDeveloperMode,
  dispatchSetDeveloperMode: setDeveloperMode,
};

const ConnectedComponent = connect(
  mapStateToProps,
  mapDispatchToProps
)(DeveloperSwitch);

const StyledComponent = withStyles(styles)(ConnectedComponent);

const TranslatedComponent = withTranslation()(StyledComponent);

export default TranslatedComponent;
