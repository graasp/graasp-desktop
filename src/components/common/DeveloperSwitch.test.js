import React from 'react';
import { shallow } from 'enzyme';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import { DeveloperSwitch } from './DeveloperSwitch';
import Loader from './Loader';

const createDeveloperSwitchProps = (developerMode, activity = false) => {
  return {
    developerMode,
    activity,
    t: text => text,
    dispatchGetDeveloperMode: jest.fn(),
    dispatchSetDeveloperMode: jest.fn(),
    classes: {
      formControl: '',
    },
  };
};

const developerModeValues = [false, true];

describe('<DeveloperSwitch />', () => {
  describe.each(developerModeValues)(
    '<DeveloperSwitch /> renders',
    developerMode => {
      let wrapper;

      beforeAll(() => {
        const props = createDeveloperSwitchProps(developerMode);
        // eslint-disable-next-line react/jsx-props-no-spreading
        wrapper = shallow(<DeveloperSwitch {...props} />);
      });

      it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
      });

      it('renders one <FormControlLabel /> component with text "Developer Mode" and right value', () => {
        const formControlLabel = wrapper.find(FormControlLabel);
        expect(formControlLabel).toHaveLength(1);
        expect(formControlLabel.prop('label')).toEqual('Developer Mode');
        expect(formControlLabel.prop('control').props.value).toBe(
          developerMode
        );
      });
    }
  );

  describe.each(developerModeValues)(
    '<DeveloperSwitch /> on activity',
    developerMode => {
      let wrapper;

      it('renders <Loader />', () => {
        const props = createDeveloperSwitchProps(developerMode, true);
        // eslint-disable-next-line react/jsx-props-no-spreading
        wrapper = shallow(<DeveloperSwitch {...props} />);
        expect(wrapper.find(Loader)).toHaveLength(1);
      });
    }
  );
});
