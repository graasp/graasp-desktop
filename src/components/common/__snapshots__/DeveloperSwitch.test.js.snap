// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DeveloperSwitch /> <DeveloperSwitch /> renders renders correctly 1`] = `
<WithStyles(ForwardRef(FormControl))
  className=""
>
  <WithStyles(ForwardRef(FormControlLabel))
    control={
      <ForwardRef(WithStyles)
        checked={false}
        color="primary"
        onChange={[Function]}
        value={false}
      />
    }
    label="Developer Mode"
  />
</WithStyles(ForwardRef(FormControl))>
`;

exports[`<DeveloperSwitch /> <DeveloperSwitch /> renders renders correctly 2`] = `
<WithStyles(ForwardRef(FormControl))
  className=""
>
  <WithStyles(ForwardRef(FormControlLabel))
    control={
      <ForwardRef(WithStyles)
        checked={true}
        color="primary"
        onChange={[Function]}
        value={true}
      />
    }
    label="Developer Mode"
  />
</WithStyles(ForwardRef(FormControl))>
`;
